src/PowerBI/PowerBIUtils.js

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Cosmo Tech.
+// Licensed under the MIT license.
+
+import { PowerBIReportEmbedSimpleFilter, PowerBIReportEmbedMultipleFilter } from './PowerBIReportEmbedFilter';
+import { ScenarioDTO } from "./ScenarioDTO";
+
+function constructDynamicValue (filterValue, obj) {
+  if (filterValue === undefined) {
+    throw new Error('value path is undefined');
+  }
+  const res = filterValue.split('.').reduce(function (o, k) {
+    return o && o[k];
+  }, obj);
+  if (res === undefined) {
+    console.log(filterValue + ' is not a valid path!!! Please adapt the configuration');
+  }
+  return res;
+}
+
+const constructDynamicFilters = (filtersConfig, obj) => {
+  const result = [];
+  for (const filterConfig of filtersConfig) {
+    const filterValues = filterConfig.values;
+    let filter;
+    if (Array.isArray(filterValues)) {
+      const values = [];
+      for (const filterValue of filterValues) {
+        const value = constructDynamicValue(filterValue, obj, filterConfig);
+        if(value !== undefined){
+          values.push(value);
+        }
+      }
+      if(values.length !== 0){
+        filter = new PowerBIReportEmbedMultipleFilter(filterConfig.target.table, filterConfig.target.column, values);
+        result.push(filter);
+      }
+    } else if (typeof filterValues === 'string') {
+      const filterValue = filterConfig.values;
+      const value = constructDynamicValue(filterValue, obj, filterConfig);
+      if(value !== undefined){
+        filter = new PowerBIReportEmbedSimpleFilter(filterConfig.target.table, filterConfig.target.column, [value]);
+        result.push(filter);
+      }
+    }
+  }
+  return result;
+};
+
+const constructScenarioDTO = (scenario, scenarioList) => {
+  let result;
+  if (scenario !== undefined) {
+    let masterId;
+    if (scenarioList != undefined) {
+      masterId = getMasterScenarioId(scenario,scenarioList);
```

Comment:
Did you check if we can use the **rootId** property ? We could thus remove the _getMasterScenarioId_ function and improve performance for large scenarios lists

Reply:
rootId corresponds to masterId after checks.
I change that ;)